### PR TITLE
fix: Rerun filetransfer tests on 408 timeouts

### PR DESCRIPTION
**Pull request by oleorhagen**
This is a pretty serious refactor of the filetransfer tests.

In essence this splits them up into separate tests, as opposed to having a lot
of different tests serialised in one test as before.

The class structure is used to a larger extent now, with a base test class
holding the tests, and then the sub OS and Enterprise test classes is simply
overriding the needed fixtures in order to provide the given OS and enterprise setups.

Tests have also been rewritten to be re-entrant. This was needed, since we now
rerun a given test on 408 timeouts (which we know can happen occasionally).
Hence, the names used in the tests are randomly generated, and not static and
semantic as before. However, I do not believe this should be a problem.

Doing this enables us to rerun each test a given number of times (in this
instance by flaky, so default=3).

Also, this opens the door to running these tests in parallel, even though they
are using the same client in the setup.

I want to double highlight the above. All tests in a given class (OS or
Enterprise) share a given setup environment. This means that all tests are run
with the same backend, and the same client.

Take heed!

Ticket: QA-456
Changelog: None

Signed-off-by: Ole Petter <ole.orhagen@northern.tech>

### CODE REVIEW

**Comment by oleorhagen** [comment]
Note: There is still a `WIP` test, but a review of 98% of this can still be done. The commit message should describe fairly well what is going on I hope.

I apologise if this is a bit much, but mostly it is rearranging stuff. Please have a look, and then we'll see what I can get done tomorrow :)

**Comment by oleorhagen** [comment]
@mender-test-bot start pipeline

**Comment by oleorhagen** [comment]
@mender-test-bot start pipeline

**Review by kacf — APPROVED** [review]
Looks good to me!

As we talked about, cleanup might be an alternative, or even better, an addition to this. But I understand where these tests are coming from, most of them assume the services are disposable, and it's not something we want to spend a lot of time on changing.

> Also, this opens the door to running these tests in parallel, even though they
> are using the same client in the setup.

Each xdist worker has its own instance of the services, so this would have worked regardless actually.

**Comment by oleorhagen** [comment]
> Looks good to me!
> 
> As we talked about, cleanup might be an alternative, or even better, an addition to this. But I understand where these tests are coming from, most of them assume the services are disposable, and it's not something we want to spend a lot of time on changing.
> 
> > Also, this opens the door to running these tests in parallel, even though they
> > are using the same client in the setup.
> 
> Each xdist worker has its own instance of the services, so this would have worked regardless actually.

Yes, this is true. But now there are more tests to run in parallel, so in a sense the parallel factor has increased :)

**Review by lluiscampos — APPROVED** [review]
Well done! :rocket:

File: tests/tests/test_filetransfer.py [hunk 1]
```diff
@@ -400,244 +441,265 @@ def test_filetransfer_limits_upload(self, mender_device, devid, auth):
...
 
 
-class TestFileTransfer(BaseTestFileTransfer):
+def rerun_on_timeouts(err, *args):
+    if not issubclass(err[0], AssertionError):
+        return False
+    return "408" in str(err[1])
+
+
+@flaky(rerun_filter=rerun_on_timeouts)
```
[the remark this answers is not in the document]
Comment (oleorhagen) [reply tests/tests/test_filetransfer.py:678]: :books:

[the remark this answers is not in the document]
Comment (oleorhagen) [reply tests/tests/test_filetransfer.py:678]: :smile_cat:

**Comment by oleorhagen** [comment]
@mender-test-bot start pipeline

**Comment by oleorhagen** [comment]
I have now run this locally for a long while, and many times, and the spurious failures I've seen seem to be gone.

tldr; I'm now confident enough that this can be merged and cause more good than harm.

Merging once I get a green pipeline.
